From 5116ebcb25de54a6987c5e3b0a958b9d8a471532 Mon Sep 17 00:00:00 2001
From: Robin Snyders <robin@snyders.xyz>
Date: Fri, 28 Aug 2026 18:57:05 +0200
Subject: [PATCH 2/2] ASoC: qcom: q6afe-dai: configure DisplayPort before start

Qualcomm downstream sends the MST stream and DPTX controller indices
before opening the shared AFE port. Configure the only DisplayPort path
exposed by q6afe as controller zero, stream zero.

Defer the HDMI and DisplayPort AFE start to trigger so hdmi-codec has
enabled the controller audio engine first. Clear the local started flag
when prepare stops an existing port.

Signed-off-by: Robin Snyders <robin@snyders.xyz>
---
 sound/soc/qcom/qdsp6/q6afe-dai.c | 49 ++++++++++++++++++++++++++++++--
 1 file changed, 46 insertions(+), 3 deletions(-)

diff --git a/sound/soc/qcom/qdsp6/q6afe-dai.c b/sound/soc/qcom/qdsp6/q6afe-dai.c
index aa9ca534a81f2..0548d4c6ed5e2 100644
--- a/sound/soc/qcom/qdsp6/q6afe-dai.c
+++ b/sound/soc/qcom/qdsp6/q6afe-dai.c
@@ -397,14 +397,13 @@ static int q6afe_dai_prepare(struct snd_pcm_substream *substream,
 			dev_err(dai->dev, "fail to close AFE port (%d)\n", rc);
 			return rc;
 		}
+		dai_data->is_port_started[dai->id] = false;
 	}
 
 	switch (dai->id) {
 	case HDMI_RX:
 	case DISPLAY_PORT_RX:
-		q6afe_hdmi_port_prepare(dai_data->port[dai->id],
-					&dai_data->port_config[dai->id].hdmi);
-		break;
+		return 0;
 	case SLIMBUS_0_RX ... SLIMBUS_6_TX:
 		q6afe_slim_port_prepare(dai_data->port[dai->id],
 					&dai_data->port_config[dai->id].slim);
@@ -706,10 +705,54 @@ static const struct snd_soc_dai_ops q6afe_usb_ops = {
 	 */
 };
 
+static int q6hdmi_trigger(struct snd_pcm_substream *substream, int cmd,
+			  struct snd_soc_dai *dai)
+{
+	struct q6afe_dai_data *dai_data = dev_get_drvdata(dai->dev);
+	struct q6afe_port *port = dai_data->port[dai->id];
+	int rc;
+
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_START:
+	case SNDRV_PCM_TRIGGER_RESUME:
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+		if (dai_data->is_port_started[dai->id])
+			return 0;
+
+		if (dai->id == DISPLAY_PORT_RX) {
+			rc = q6afe_port_set_display_stream(port, 0, 0);
+			if (rc < 0) {
+				dev_err(dai->dev,
+					"fail to configure AFE display stream (%d)\n",
+					rc);
+				return rc;
+			}
+		}
+
+		q6afe_hdmi_port_prepare(port,
+					&dai_data->port_config[dai->id].hdmi);
+
+		rc = q6afe_port_start(port);
+		if (rc < 0) {
+			dev_err(dai->dev, "fail to start AFE port %x (%d)\n",
+				dai->id, rc);
+			return rc;
+		}
+
+		dai_data->is_port_started[dai->id] = true;
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
 static const struct snd_soc_dai_ops q6hdmi_ops = {
 	.probe			= msm_dai_q6_dai_probe,
 	.remove			= msm_dai_q6_dai_remove,
 	.prepare		= q6afe_dai_prepare,
+	.trigger		= q6hdmi_trigger,
 	.hw_params		= q6hdmi_hw_params,
 	.shutdown		= q6afe_dai_shutdown,
 };
